src/libraries/Microsoft.PowerFx.Core/IR/Visitors/RewritingIRVisitor.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core.IR.Nodes;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.IR
+{
+    // Common helper for IR Rewriting. 
+    // IR trees are immutable, so we need to walk and rewrite the tree to include any new nodes.
+    // Only return new nodes if needed, so if there are no changes, the original object instances are returned.
+    internal abstract class RewritingIRVisitor<TRet, TCtx> : IRNodeVisitor<TRet, TCtx>
+    {
+        protected abstract IntermediateNode Materialize(TRet ret);
+
+        protected abstract TRet Ret(IntermediateNode node);
+
+        public override TRet Visit(TextLiteralNode node, TCtx context)
```

Comment:
/// <summary> comments for these methods would be helpful for future implementers.